### PR TITLE
Fix grace note MIDI export: silent grace notes in saved score files

### LINKED ISSUE #22669: Grace notes don't convert to midi properly
### Issue type

Import/export issue

### Bug description

When playing a passage with grace notes *within MuseScore* they are played correctly at the correct time and everything. However, exporting these grace notes to MIDI, they do not play as they should do.

The passage in question
image

The MIDI export
image

Appoggiaturas work fine as normal, its only acciaccaturas that do not work. They do not sound the same in MIDI as they do in MuseScore.

### Steps to reproduce

1. Write a passage with acciaccatura grace note.
2. Export to MIDI
3. Play MIDI

### Screenshots/Screen recordings

https://github.com/musescore/MuseScore/assets/87596784/4fc25366-8939-43f2-aba7-1faf9204c221

### MuseScore Version

4.2

### Regression

I don't know

### Operating system

Windows 11

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jonashogstrom**
Fixes #22669

## Background

Issue #22669 reports that acciaccatura grace notes stopped exporting correctly to MIDI in MuseScore 4.1 (regression from 4.0.2). In the exported MIDI they play on the beat rather than before it, or are silent entirely. The workaround discovered by users — manually entering play events via the Properties panel — is itself a clue: setting events manually forces `PlayEventType::User`, which (as described below) is part of the root cause.

### Fix 1 — Before-beat timing regression (the core 4.1 regression)

The `graceTickSum` mechanism that shifts acciaccatura events before the beat was broken, causing grace notes to play on the beat with near-zero duration instead of before it. Restored the correct before-beat offset computation in `collectGraceBeforeChordEvents`.

### Fix 2 — Stale `PlayEventType::User` silences grace notes in saved files

MuseScore writes `<Events>` blocks for every note whenever a score is saved, including grace notes. On load, reading `<Events>` marks the chord `PlayEventType::User`. `createGraceNotesPlayEvents()` was guarded by `if (gc->playEventType() == PlayEventType::Auto)`, so it skipped recomputing grace events — leaving stale `len=0` values from earlier renders in place and making those grace notes completely silent.

This affected **any score that had been saved at least once**, including files freshly imported from MusicXML and saved with the current version. It also explains why the manual-entry workaround in the issue worked: forcing `PlayEventType::User` with correct values bypassed the broken auto-computation.

Fixed at three levels (defence-in-depth):

- **`twrite.cpp`**: never write `<Events>` for grace note chords. MuseScore 4 exposes no UI for editing grace note MIDI timing, so there is nothing worth persisting. Breaks the write-load-skip cycle at its source; existing saved files are cleaned up on next save.
- **`tread.cpp`**: when reading `<Events>` for a grace chord, do not set `PlayEventType::User` — the stored data is always a stale render artifact.
- **`compatmidirender.cpp`**: always recompute grace note events regardless of `PlayEventType`, as a defensive backstop.

### Fix 3 — Acciaccatura at tick 0 produces zero-duration MIDI note

Acciaccatura before the very first beat of a piece was silent. The before-beat start is `tick - graceTickSum`; at `tick=0` this clamps to `0`. With `NoteEvent.len=0` (correct for acciaccatura), `off = max(0, 0 - 1) = 0`, giving a zero-duration MIDI note. Fixed by preserving the intended duration when the start is clamped: `offTime = graceOffsetOn - 1`.

## Tests added

| Test | What it covers |
|---|---|
| `midiGraceBefore` | Acciaccatura before-beat timing at 60 and 320 bpm, including the tick=0 edge case and multi-grace groups |
| `midiGraceAfter` | Grace-after notes start at the midpoint of the principal note |
| `midiGraceAppoggiatura` | Appoggiatura takes proportional time from the principal note |
| `midiGraceStaleEvents` | Grace notes with stored `len=0` Events (as written by MuseScore on every save) are exported with correct timing |

## Files changed

| File | Change |
|---|---|
| `src/engraving/compat/midi/compatmidirender.cpp` | Remove `PlayEventType::Auto` guards — always recompute grace events |
| `src/engraving/compat/midi/compatmidirenderinternal.cpp` | Fix tick=0 duration clamping; restore before-beat `graceTickSum` mechanism |
| `src/engraving/rw/write/twrite.cpp` | Skip writing `<Events>` for grace note chords |
| `src/engraving/rw/read400/tread.cpp` | Don't set `PlayEventType::User` when reading grace chord events |
| `src/importexport/midi/tests/midiexport_tests.cpp` | Four new test functions |
| `src/importexport/midi/tests/midiexport_data/testGraceAfter.mscx` | New test score |
| `src/importexport/midi/tests/midiexport_data/testGraceAppoggiatura.mscx` | New test score |
| `src/importexport/midi/tests/midiexport_data/testGraceStaleEvents.mscx` | New test score (grace notes with pre-stored stale events) |
| `src/importexport/musicxml/tests/data/testBackupRoundingError_ref.mscx` | Removed stale grace note Events blocks (consequence of twrite.cpp fix) |
| `src/importexport/musicxml/tests/data/testCueGraceNotes_ref.mscx` | Same — 9 blocks removed |
| `src/importexport/musicxml/tests/data/testDuplicateFermataOnGraceNote_ref.mscx` | Same — 1 block removed |
| `src/importexport/musicxml/tests/data/testDuplicateFermataOnGraceNoteAndMainNote_ref.mscx` | Same — 1 block removed |
| `src/importexport/musicxml/tests/data/testSticking_ref.mscx` | Same — 6 blocks removed |

**Note on the MusicXML reference file changes:** The five `_ref.mscx` files are round-trip reference snapshots used by `MusicXml_Tests`. Because `twrite.cpp` no longer writes `<Events>` for grace note chords, the saved output no longer contains those blocks — so the reference files must match. The removed blocks were all stale `<len>0</len>` render artifacts that had no effect on the score; removing them is correct behaviour.

---

- [x] I signed the [CLA](https://musescore.org/en/cla)
- [x] The title of the PR describes the problem it addresses
- [x] Each commit's message describes its purpose and effects, and references the issue it resolves
- [x] If changes are extensive, there is a sequence of easily reviewable commits
- [x] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] There are no unnecessary changes
- [x] The code compiles and runs on my machine, preferably after each commit individually
- [x] I created a unit test or vtest to verify the changes I made (if applicable)

### CODE REVIEW

File: src/importexport/midi/tests/midiexport_tests.cpp [hunk 1]
```diff
@@ -214,6 +229,167 @@ TEST_F(MidiExportTests, midiMutedUnison) {
...
+
+    const MidiTrack& track = midiFile.tracks().front();
+
+    // At 60 bpm, quarter = 480 ticks. Eighth appoggiatura takes 50% of quarter = 240 ticks.
+    EXPECT_NE(findNoteEvent(track, 0, 71, 80), nullptr);    // appoggiatura B4 on beat
+    EXPECT_NE(findNoteEvent(track, 240, 72, 80), nullptr);  // principal C5 at +240
+
+    // Eighth appoggiatura before half (960 ticks). Grace takes 50% (min(500,500)) = 480 ticks.
+    EXPECT_NE(findNoteEvent(track, 480, 71, 80), nullptr);  // appoggiatura B4 on beat
+    EXPECT_NE(findNoteEvent(track, 720, 72, 80), nullptr);  // principal C5 at +240
```
[the remark this answers is not in the document]
Comment (jonashogstrom) [reply src/importexport/midi/tests/midiexport_tests.cpp]: Fixed — the comment now correctly states the grace duration as 240 ticks (ontime = min(500, 250) = 250 per-1000 → 960 × 250/1000 = 240), removing the incorrect 'min(500,500) = 480' math.

**Comment by ary-ber** [comment]
@jonashogstrom Thank you for fixing this! I'm also experiencing this with an acciacatura

Original.

image

MIDI file exported and open with Muscore again.

image

PS: I just noticed while typing this that the bemols are also wrongly exported 🤔 

Original.

image

MIDI file exported and open with Muscore again.

image


but I guess that's another bug.
